sandbox/apparmor/apparmor.go

```diff
@@ -329,46 +335,106 @@ func probeKernelFeatures() ([]string, error) {
 }
 
 func probeParserFeatures() ([]string, error) {
-	parser, err := findAppArmorParser()
+	parser, args, internal, err := FindAppArmorParser()
 	if err != nil {
 		return []string{}, err
 	}
-	features := make([]string, 0, 4)
-	if tryAppArmorParserFeature(parser, "change_profile unsafe /**,") {
-		features = append(features, "unsafe")
-	}
-	if tryAppArmorParserFeature(parser, "network qipcrtr dgram,") {
-		features = append(features, "qipcrtr-socket")
-	}
-	if tryAppArmorParserFeature(parser, "capability bpf,") {
-		features = append(features, "cap-bpf")
-	}
-	if tryAppArmorParserFeature(parser, "capability audit_read,") {
-		features = append(features, "cap-audit-read")
+	var featureProbes = []struct {
+		feature string
+		probe   string
+	}{
+		{
+			feature: "unsafe",
+			probe:   "change_profile unsafe /**,",
+		},
+		{
+			feature: "include-if-exists",
+			probe:   `#include if exists "/foo"`,
+		},
+		{
+			feature: "qipcrtr-socket",
+			probe:   "network qipcrtr dgram,",
+		},
+		{
+			feature: "mqueue",
+			probe:   "mqueue,",
+		},
+		{
+			feature: "cap-bpf",
+			probe:   "capability bpf,",
+		},
+		{
+			feature: "cap-audit-read",
+			probe:   "capability audit_read,",
+		},
+	}
+	features := make([]string, 0, len(featureProbes)+1)
+	for _, fp := range featureProbes {
+		if tryAppArmorParserFeature(parser, args, fp.probe) {
+			features = append(features, fp.feature)
+		}
 	}
-	if tryAppArmorParserFeature(parser, "mqueue,") {
-		features = append(features, "mqueue")
+	if internal {
+		features = append(features, "snapd-internal")
 	}
 	sort.Strings(features)
 	return features, nil
 }
 
-// findAppArmorParser returns the path of the apparmor_parser binary if one is found.
-func findAppArmorParser() (string, error) {
+func snapdAppArmorSupportsReexecImpl() bool {
+	hostInfoDir := filepath.Join(dirs.GlobalRootDir, dirs.CoreLibExecDir)
+	_, flags, err := snapdtool.SnapdVersionFromInfoFile(hostInfoDir)
+	return err == nil && flags["SNAPD_APPARMOR_REEXEC"] == "1"
+}
```

Comment:
this needs a standalone test using dirs.SetRootDir and possibly export_test.go

Reply:
Added in f4c6caae95